### PR TITLE
Address incomplete patterns

### PR DESCRIPTION
**Pull request by jacereda**

### CODE REVIEW

**Comment by jacereda** [comment]
This is a first step, IMHO we should try to follow the 'make illegal states unrepresentable' path. Ideally we should get rid of most of these error paths over time by providing strong warrantees (by providing `NonEmpty` lists to commands, for example).

**Comment by jacereda** [comment]
The horrible error messages are deliberate, so we have some motivation to get rid of them ;)

File: src/Deftype.hs [hunk 1]
```diff
@@ -428,6 +434,7 @@ binderForStrOrPrn typeEnv env insidePath structTy@(StructTy (ConcreteNameTy type
             (concreteStr typeEnv env structTy (memberXObjsToPairs membersXObjs) strOrPrn)
             ("converts a `" ++ typeName ++ "` to a string.")
         )
+binderForStrOrPrn _ _ _ _ _ _ = error "binderforstrorprn"
```
Comment (eriksvedang) [inline src/Deftype.hs:437]: I want to see the face of a person who gets "binderforstrorprn" as an error message :D

**Review by eriksvedang — APPROVED** [review]
Really nice, I see it has already caught a couple of lurking bugs / mistakes!

**Comment by eriksvedang** [comment]
> The horrible error messages are deliberate, so we have some motivation to get rid of them ;)

I like your thinking!

**Comment by eriksvedang** [comment]
> This is a first step, IMHO we should try to follow the 'make illegal states unrepresentable' path. Ideally we should get rid of most of these error paths over time by providing strong warrantees (by providing `NonEmpty` lists to commands, for example).

Yes, please – sounds awesome!

**Pull request merged**